Userland/Libraries/LibWeb/Layout/TreeBuilder.cpp

```diff
@@ -671,6 +675,59 @@ void TreeBuilder::generate_missing_parents(NodeWithStyle& root)
         else
             parent.append_child(*wrapper);
     }
+
+    return table_roots_to_wrap;
+}
+
+template<typename Matcher, typename Callback>
+static void for_each_child_box_matching(Box& parent, Matcher matcher, Callback callback)
+{
+    parent.for_each_child_of_type<Box>([&](Box& child_box) {
+        if (matcher(child_box))
+            callback(child_box);
+    });
 }
 
+static void fixup_row(Box& row_box, TableGrid const& table_grid, size_t row_index)
+{
+    bool missing_cells_run_has_started = false;
+    for (size_t column_index = 0; column_index < table_grid.column_count(); ++column_index) {
+        if (table_grid.occupancy_grid().contains({ column_index, row_index })) {
+            VERIFY(!missing_cells_run_has_started);
+            continue;
+        }
+        missing_cells_run_has_started = true;
+        auto cell_element = row_box.heap().template allocate<HTML::HTMLTableCellElement>(
+            row_box.document().realm(),
+            row_box.document(),
+            DOM::QualifiedName { HTML::TagNames::td, ""sv, Namespace::HTML });
+        cell_element->set_is_generated();
+        MUST(row_box.dom_node()->append_child(cell_element)); // FIXME propagate errors
```

Comment:
fixup should not affect DOM tree. missing cells are supposed to be added only in layout tree.

Reply:
I'm not sure I can add cells without a DOM node. I'll give it another try, but I remember it crashing.

Reply:
Ok, this worked fine without DOM nodes.